services/port_manager/src/main/java/com/futurewei/alcor/portmanager/restwrap/IpAddressRestWrap.java

```diff
@@ -0,0 +1,154 @@
+import com.futurewei.alcor.portmanager.rollback.AbstractIpAddrRollback;
+import com.futurewei.alcor.portmanager.utils.BeanUtil;
+import com.futurewei.alcor.web.entity.ip.IpAddrRequest;
+import com.futurewei.alcor.web.entity.ip.IpVersion;
+import com.futurewei.alcor.web.entity.port.PortState;
+import com.futurewei.alcor.web.entity.SubnetStateJson;
+import com.futurewei.alcor.web.rest.IpAddressRest;
+import com.futurewei.alcor.portmanager.exception.IpAddrInvalidException;
+import com.futurewei.alcor.portmanager.exception.IpVersionInvalidException;
+import com.futurewei.alcor.portmanager.rollback.AllocateIpAddrRollback;
+import com.futurewei.alcor.portmanager.rollback.PortStateRollback;
+import com.futurewei.alcor.portmanager.rollback.ReleaseIpAddrRollback;
+import com.futurewei.alcor.web.rest.SubnetRest;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+
+public class IpAddressRestWrap {
+    private IpAddressRest ipAddressRest;
+    private SubnetRest subnetRest;
+    private Stack<PortStateRollback> rollbacks;
+    private String projectId;
+
+    public IpAddressRestWrap(Stack<PortStateRollback> rollbacks, String projectId) {
+        ipAddressRest = BeanUtil.getBean(IpAddressRest.class);
+        subnetRest = BeanUtil.getBean(SubnetRest.class);
+        this.rollbacks = rollbacks;
+        this.projectId = projectId;
+    }
+
+    private int getIpVersion(String ipAddress) throws Exception {
+        if (Ipv4AddrUtil.formatCheck(ipAddress)) {
+            return IpVersion.IPV4.getVersion();
+        } else if (Ipv6AddrUtil.formatCheck(ipAddress)) {
+            return IpVersion.IPV6.getVersion();
+        } else {
+            throw new IpAddrInvalidException();
+        }
+    }
+
+    private String getRangeIdBySubnetId(String subnetId, int ipVersion) throws Exception {
+        SubnetStateJson subnetStateJson = subnetRest.getSubnetState(projectId, subnetId);
+        if (subnetStateJson == null || subnetStateJson.getSubnet() == null) {
+            throw new VerifySubnetException();
+        }
+
+        if (IpVersion.IPV4.getVersion() == ipVersion) {
+            return subnetStateJson.getSubnet().getIpV4RangeId();
+        } else if (IpVersion.IPV6.getVersion() == ipVersion) {
+            return subnetStateJson.getSubnet().getIpV4RangeId();
+        }
+
+        throw new IpVersionInvalidException();
+    }
+
+    private void addIpAddrRollback(AbstractIpAddrRollback rollback, IpAddrRequest ipAddr) {
+        if (rollback instanceof AllocateIpAddrRollback) {
+            rollback.putAllocatedIpAddress(ipAddr);
+        } else {
+            rollback.putReleasedIpAddress(ipAddr);
+        }
+
+        rollbacks.push(rollback);
+    }
+
+    public List<IpAddrRequest> allocateIpAddress(Object args) throws Exception {
+        List<IpAddrRequest> ipAddrRequests = new ArrayList<>();
+        PortState portState = (PortState)args;
+
+        IpAddrRequest result = ipAddressRest.allocateIpAddress(IpVersion.IPV4,
```

Comment:
How about Ipv6? Could we use sth like "portState.IpVersion" instead of fixing with IPv4?

Reply:
There is no ipVersion field in portState, do you think only ipv4 address is assigned to port by default, or both ipv4 and ipv6 addresses are assigned?

Reply:
I see. Ipv4v6 is something we need to support as well. Maybe put it in your backlog and do it when you get a chance.